### PR TITLE
Check reference at snapshot creation

### PR DESCRIPTION
**Pull request by jordanschalm**
This PR extends https://github.com/onflow/flow-go/pull/4086, adding checks to snapshot creation to differentiate between exceptions and unknown blocks in BlockSignerDecoder. 
- Updates `protocol.Snapshot` to guarantee that `ErrUnknownSnapshotReference` is returned in all cases where a snapshot's reference is unknown
  - Add existence checking methods to `storage/common`, `storage.Cache`
- Updates BlockSignerDecoder to differentiate between errors from an unknown input block and exception
- Misc: marks all unexpected errors in `storage/common` as exceptions
- Misc: update `flow-emu` version to include https://github.com/onflow/flow-emulator/pull/280 (avoid changes to `Headers` being breaking)

### CODE REVIEW

File: state/protocol/badger/state.go [hunk 1]
```diff
@@ -622,18 +632,30 @@ func (state *State) AtHeight(height uint64) protocol.Snapshot {
...
+// This function may return an invalid.Snapshot with:
+//   - state.ErrUnknownSnapshotReference:
+//     -> if no block with the given ID exists in the state
+//   - exception for critical unexpected storage errors
 func (state *State) AtBlockID(blockID flow.Identifier) protocol.Snapshot {
-	// TODO should return invalid.NewSnapshot(ErrUnknownSnapshotReference) if block doesn't exist
-	return NewSnapshot(state, blockID)
+	exists, err := state.headers.Exists(blockID)
+	if err != nil {
+		return invalid.NewSnapshotf("could not check existence of reference block")
```
Comment (peterargue) [inline state/protocol/badger/state.go:647]: should this include the error message for context?

**Review by peterargue — APPROVED** [review]

File: state/protocol/badger/state.go (hunk 1 above)
Comment (jordanschalm) [inline state/protocol/badger/state.go:647]:
```suggestion
		return invalid.NewSnapshotf("could not check existence of reference block: %w", err)
```

Comment (jordanschalm) [reply state/protocol/badger/state.go]: https://github.com/onflow/flow-go/pull/4093/commits/b81799f88d0c3ca17632a40fd888728edead45c4

File: consensus/hotstuff/signature/block_signer_decoder_test.go [hunk 2]
```diff
@@ -65,31 +65,57 @@ func (s *blockSignerDecoderSuite) Test_RootBlock() {
...
+	s.Run("ByEpoch exception", func() {
+		exception := errors.New("unexpected exception")
+		*s.committee = *hotstuff.NewDynamicCommittee(s.T())
+		s.committee.On("IdentitiesByEpoch", mock.Anything).Return(nil, exception)
+
+		ids, err := s.decoder.DecodeSignerIDs(s.block.Header)
+		require.Empty(s.T(), ids)
+		require.NotErrorIs(s.T(), err, model.ErrViewForUnknownEpoch)
+		require.False(s.T(), signature.IsInvalidSignerIndicesError(err))
+		require.True(s.T(), errors.Is(err, exception))
```
Comment (durkmurder) [inline consensus/hotstuff/signature/block_signer_decoder_test.go:81]:
```suggestion
		require.ErrorIs(s.T(), err, exception)
```

File: consensus/hotstuff/signature/block_signer_decoder_test.go [hunk 3]
```diff
@@ -65,31 +65,57 @@ func (s *blockSignerDecoderSuite) Test_RootBlock() {
...
+		exception := errors.New("unexpected exception")
+		*s.committee = *hotstuff.NewDynamicCommittee(s.T())
+		s.committee.On("IdentitiesByEpoch", mock.Anything).Return(nil, model.ErrViewForUnknownEpoch)
+		s.committee.On("IdentitiesByBlock", mock.Anything).Return(nil, exception)
+
+		ids, err := s.decoder.DecodeSignerIDs(s.block.Header)
+		require.Empty(s.T(), ids)
+		require.NotErrorIs(s.T(), err, model.ErrViewForUnknownEpoch)
+		require.False(s.T(), signature.IsInvalidSignerIndicesError(err))
+		require.True(s.T(), errors.Is(err, exception))
```
Comment (durkmurder) [inline consensus/hotstuff/signature/block_signer_decoder_test.go:93]:
```suggestion
		require.ErrorIs(s.T(), err, exception)
```

File: storage/badger/operation/common_test.go [hunk 4]
```diff
@@ -270,7 +271,45 @@ func TestRetrieveUnencodeable(t *testing.T) {
 
 		var act *UnencodeableEntity
 		err := db.View(retrieve(key, &act))
-		require.ErrorIs(t, err, errCantDecode)
+		require.Error(t, err)
+		require.NotErrorIs(t, err, storage.ErrNotFound)
+	})
+}
+
+func TestExists(t *testing.T) {
```
Comment (durkmurder) [inline storage/badger/operation/common_test.go:279]:
```suggestion
// TestExists verifies that `exists` returns correct results in different scenarios.
func TestExists(t *testing.T) {
```

**Comment by jordanschalm** [comment]
bors merge

**Comment by jordanschalm** [comment]
```
2023-04-04T15:51:44.6748785Z ··· DOCK: 2023-04-04 15:51:44.674247022 +0000 UTC (clogs ) consensus_test_3          (d4ff9c313741f4d44cbd69fe1413440a1b94928488b0987fddf76152c7274477) - ������{"level":"fatal","node_role":"consensus","node_id":"4845cba675661fdf0a56902be92d82bca101863e94cecadc662632f699a9ca70","error":"internal error processing queued message: processing collection guarantee unexpected err: origin validation error: unexpected error checking collection origin 5fd78b82640cd6e2b566a24f7b31e397ffd3814c504ad7eb30453dfd90cdf6d9 at reference block ca1f2fbd92c7b7eccc0c0873ae42fbdaea39137605a8ef4c602d45fca086fcd1: could not retrieve node identity (id=5fd78b82640cd6e2b566a24f7b31e397ffd3814c504ad7eb30453dfd90cdf6d9): unknown block ca1f2fbd92c7b7eccc0c0873ae42fbdaea39137605a8ef4c602d45fca086fcd1: reference block of the snapshot is not resolvable)","time":"2023-04-04T15:51:44.6738984Z","message":"unhandled irrecoverable error"}
```

The integration tests failed because some calls had not been updated to expect `ErrUnknownSnapshotReference` - addressed in https://github.com/onflow/flow-go/pull/4093/commits/2fa9fc5cee3494d710ff1b9bd0521002fbd0627b

**Comment by jordanschalm** [comment]
bors merge

**Comment by jordanschalm** [comment]
bors merge

**Pull request merged**